This PR is a tool to see relevant changes contained in the incremental Unit Test PR.

macrogreg opened the pull request:
The PR https://github.com/temporalio/sdk-dotnet/pull/53 is incremental to the previous PR, and so it includes changes from all previous PRs that are not yet merged. To make the review easier, this PR is a diff between https://github.com/temporalio/sdk-dotnet/pull/53 and the branch it is based on (i.e, the PR immediately preceding). Thus, only relevant delta is included.

mightyshazam committed:
Added unit tests to cover Temporal.Sdk.Common

mightyshazam committed:
added unit tests again with autocrlf enabled

mightyshazam committed:
Merge branch 'master' into feature/unit-tests-2

mightyshazam committed:
initial attempt at ci build

mightyshazam committed:
updated ci to test this branch

mightyshazam committed:
updated paths

mightyshazam committed:
trying to get build paths correct

mightyshazam committed:
added framework arguments for ci builds

mightyshazam committed:
updated path on repository checkout

mightyshazam committed:
corrected spelling of netcoreapp target

mightyshazam committed:
removed windows and macos targets

mightyshazam committed:
updated nuget restore path

mightyshazam committed:
updated msbuild args

mightyshazam committed:
renamed ci file and updated msbuild commands

mightyshazam committed:
updated restore target

mightyshazam committed:
removed system.text.json for short term compatibility

mightyshazam committed:
Merge remote-tracking branch 'macrogreg/Client07-DataConverters' into feature/unit-tests-2

mightyshazam committed:
Merge remote-tracking branch 'macrogreg/Client09-StyleCheckers' into feature/unit-tests-2

mightyshazam committed:
removed var

mightyshazam committed:
fixed formatting issues in tests

mightyshazam committed:
added tests for new converters

macrogreg commented on Src/SDK/Common/Temporal.Common.Payloads/internal/PayloadContainersEnumerableJsonConverter.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace Temporal.Common.Payloads
+{
+    public class PayloadContainersEnumerableJsonConverter : JsonConverter<PayloadContainers.Enumerable>
```
Why is this class necessary? The default conversion for `PayloadContainers.Enumerable` has worked fine so far. But I may be missing a scenario. Can you please elaborate?

If we need this custom converter, why is this type public? If it needs to be public, let's move it into the public directory in this namespace's folder. However, unless there is a need, I would prefer to keep it internal.

mightyshazam replied:
My unit test failed until I wrote the custom converter. The data wouldn't serialize. I'm glad you brought it up because I don't want it to be necessary, so the problem could be the assumptions I made when I wrote the test.

macrogreg replied:
Got it. We can discuss and potentially investigate together.

macrogreg replied:
Found the underlying issue and resolved the issue by adding proper treatment for JObject.

macrogreg commented on Src/SDK/Common/Temporal.Common.Payloads/public/PayloadContainers.Enumerable.cs:
```diff
@@ -1,12 +1,14 @@
 {
     public static partial class PayloadContainers
     {
+        [JsonConverter(typeof(PayloadContainersEnumerableJsonConverter))]
```
In which scenario is the existing default conversion not sufficient?

mightyshazam replied:
It didn't work in the unit test I created, so that is why I added the converter. If the assumptions in the test are wrong, then this is unnecessary. The tests, I believe, were for the Newtonsoft json converter.

macrogreg replied:
Yes, Newtonsoft. Let's chat on Monday and figure it out.

macrogreg replied:
Found the underlying issue and resolved the issue by adding proper treatment for JObject. JsonConverter no longer needed and removed.

macrogreg commented on Src/SDK/Common/Temporal.Common.Payloads/public/PayloadContainers.Enumerable.cs:
```diff
@@ -72,6 +74,7 @@ public virtual IEnumerator GetEnumerator()
             }
         }
 
+        [JsonConverter(typeof(PayloadContainersEnumerableJsonConverter<>))]
```
Same :)

macrogreg replied:
✔️

macrogreg commented on Src/SDK/Common/Temporal.Common.Payloads/public/PayloadContainers.Unnamed.Empty.cs:
```diff
@@ -25,7 +25,8 @@ public TVal GetValue<TVal>(int index)
                 public bool TryGetValue<TVal>(int index, out TVal value)
                 {
-                    throw CreateNoSuchIndexException(index);
+                    value = default(TVal);
```
In all 3 of the iface implementations I was going for applying the Try-logic to the validity of the specified type conversion, and for throwing in case the index is bad.
However, I think you are right: it is more in the spirit of the Try pattern (and thus user expectations) to apply the Try-logic wholistically and also return false rather than throw in case of bad index. :)
Thanks for picking this up.

macrogreg replied:
Same for the equivalent change in other implementations of the iface. :)

mightyshazam replied:
I made this change because try methods shouldn't throw except in the exceptional cases. An index out of bounds is the case we are handling here, so that is why I thought it shouldn't throw. However, there is another class where I left the exception in TryGetValue. It was because type deserialization fails. I also don't think that method should throw an exception, but without a huge refactor to make TryGetValue return why it didn't work, I think throwing an exception is the best way to go.

macrogreg replied:
agreed. :)

btw,. where did you need to leave the exception?

mightyshazam replied:
I left the exception in PayloadContainers.Unnamed.SerialiazedDataBacked.TryGetValue 
https://github.com/mightyshazam/sdk-dotnet/blob/9da8848d6c70e056c7c1f9126240f48672164f31/Src/SDK/Common/Temporal.Common.Payloads/public/PayloadContainers.Unnamed.SerializedDataBacked.cs#L105
I was conflicted since I don't think Try methods should throw, but it is throwing for a reason different than all other methods.

macrogreg replied:
ToDo: Probably need to make it return false. :)

macrogreg replied:
I thought about it and left the exception. To clarify, I added the following code comment:

```
// The "Try"-pattern contract of this API implies no exceptions. However, so does the 
// downstream call to `TryDeserialize`. We trust the downstream call: It should just
// return true/false. However, if something truly exceptional happens and it throws,
// we enrich that exception a clarifying message and let it bubble up.
```

macrogreg commented on Src/SDK/Common/Temporal.Sdk.Common.csproj:
```diff
@@ -9,17 +9,13 @@
   <ItemGroup>    
-    <Compile Include="$(SharedSrcBaseDir)\Temporal.Util\internal\Validate.cs"
-             Link="$(SharedSrcBaseLabel)\Temporal.Util\internal\%(Filename)%(Extension)" />
+    <Compile Include="$(SharedSrcBaseDir)\Temporal.Util\internal\Validate.cs" Link="$(SharedSrcBaseLabel)\Temporal.Util\internal\%(Filename)%(Extension)" />
```
I think, VS automatically changed the format of the lines here. It's more readable the way it was for folks who do not have super wide screens.

macrogreg commented on Src/SDK/WorkflowClient/Temporal.WorkflowClient/internal/TemporalServiceInvoker.cs:
```diff
@@ -1,1061 +1,1061 @@
-        private record HackyWorkflowChainBindingInfo(bool IsSuccess, string WorkflowChainId, string WorkflowRunId);
-    }
-}
+﻿using System;
```
Looks like this file slipped through when you adjusted the line it seems like there is a delta in every like. Could we please fix that? Also, is there any actual delta here. I looked through, but I did not spot anything.. :)

macrogreg replied:
✔️

mightyshazam committed:
added protobuf payload converter

macrogreg commented on Src/Tests/Directory.Build.props:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I like that you placed all the tests within the `Scr` folder. It helps separating from docs, release tools etc. 
How would you feel about `Test` over `Tests` for the sub-folder? I ve seen `Test` more often in the wild, but I do not mind too much.

mightyshazam replied:
I don't have a preference. I think renaming to Test is fine. I think it is best to do it in a subsequent PR after resolving other issues so the comments on this review don't get lost.

macrogreg replied:
agreed.

mightyshazam committed:
merged new style changes

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Common/TestPayload.cs:
```diff
@@ -0,0 +1,68 @@
+        [Fact]
+        public void Test_Payload_Unnamed_With_Null_Argument()
+        {
+            Assert.Throws<ArgumentNullException>(() => Payload.Unnamed(null));
```
I was considering to use https://github.com/fluentassertions/fluentassertions. What are your thoughts on that? While the actual assertion syntax that they offer if, in my opinion, nice but not a game changer, I am a big fan of their rich messages when a test fails. They can save a bunch of time investigating a failure.

Either way, I think that it is not a reason to delay this PR.
How about we discuss it later, and if we decide to go with FluentAssertions would you be OK if in some future down the line we refactored the assertions to be based on FluentAssertions?

mightyshazam replied:
If you think it will help, I see no issue with it. I'm all for clarity.

macrogreg replied:
✔️

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Common/TestPayload.cs:
```diff
@@ -0,0 +1,68 @@
+using System;
```
I suggest that within the Test folder we mimic the structure of the `Src` folder, to make it easier to find things. Does it make sense from your perspective? If so, this would need to go to:

`\src\Test\SDK\Common\Temporal.Common\public\Payload.cs`

Notably, in some future when we had the time to add a public API verificator, we can hopefully get rid of the `\public\` element.

As far as `Payload.cs` vs `TestPayload.cs`, I have no strong preference. However, this is how the view looks like right now:
![image](https://user-images.githubusercontent.com/20691812/168194966-de717e3d-b8c1-4d9c-9808-ff8284cef12c.png)

Arguably, the `xxx.Test` namespace and the path rootet in `src\Test` already convey the key message. So we can make things a bit neater by calling the class just `Payload.cs` and the respective methods:

`Unnamed_With_Null_Argument()`
`Unnamed_With_Variadic_Arguments()`
`Unnamed_With_Array_Arguments()`
and so on.

mightyshazam replied:
I didn't follow the structure under SDK because I didn't think my tests would be 1 to 1 with the classes. I expected more composition, so I laid them out according to functional area of the public API. I have no problem with changing the structure.
I prefixed everything with test because it makes navigation easier with vscode and Rider/Resharper. If I search for Payload, I'm guaranteed to get multiple results. If I search for TestPayload, I'm only going to get one. I went with the verbose names because I was hoping I could make them clear enough that seeing the test name and the failure would indicate what went wrong. However, including fluent assertions mitigates that, and I didn't consider the view from the unit test UI.

macrogreg replied:
I use VS, together we can make it convenient for both. I am generally fine with any structure, as long as it is self-explanotary and consistent. :) Let's discuss the details on Monday.

macrogreg replied:
Names discussed and adjusted.

macrogreg commented on Src/SDK/Common/Temporal.Common.Payloads/internal/PayloadContainersEnumerableJsonConverter.cs:
```diff
@@ -0,0 +1,68 @@
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
```
So far I avoided a dependency on `System.Text.Json` Why do we need it?
There are both , `Newtonsoft.Json.JsonConverter<T>`  and `System.Text.Json.Serializationctness.JsonConverter<T>`. Which one did you have in mind and why?

By the way, when I build your code locally, it broke the build for Net Fx 4.6 since that namespace is not part of that version and requires an external package.

mightyshazam replied:
I have removed System.Text.Json in the branch where I started the CI work for that reason. However, I prefer it to Newtonsoft because it is updated, and it is an order of magnitude faster for larger payloads. Unfortunately, it has issues with serializing derived classes and dictionaries, so it probably isn't the best option.

macrogreg replied:
I also prefer System.Text.Json when I serialize _my_ data. But here, we are serializing the user's data. Types that potentially come from some library and that the user may not be able to modify. Newtonsoft has better coverage for data contact attributes that may affect the way those types are intended to be serialized. That is why I picked it. If in the future System.Text.Json catches up in that space, we can switch.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Common/TestPayload.cs:
```diff
@@ -0,0 +1,68 @@
+        [Fact]
+        public void Test_Payload_Unnamed_With_Variadic_Arguments()
+        {
+            Temporal.Common.Payloads.PayloadContainers.Unnamed.InstanceBacked<object> payload = Payload.Unnamed(new object(), new object());
```
Nitpick (feel free to ignore): consider `using Temporal.Common.Payloads;` to shorten this.

mightyshazam replied:
This is the product of me using Resharper to rapidly change these from var to the type after merging in the style changes. I didn't notice.

macrogreg replied:
:)

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Common/TestPayload.cs:
```diff
@@ -0,0 +1,68 @@
+            AssertUnnamedCorrectness(length, payload);
+        }
+
+        private static void AssertUnnamedCorrectness<T>(
```
Everywhere else the project uses the following code style. My apologies, I do not know how to make the latest version of the style checker to actually flag it, so sorry for having to bring it up in the code review:

```cs
private static void AssertUnnamedCorrectness<T>(int length,
                                                Temporal.Common.Payloads.PayloadContainers.Unnamed.InstanceBacked<T> payload)
{
}
```

:)

mightyshazam replied:
Our style guide at Aware enforces it as a build error if I don't type it the way I did. It's habit. It's going to pop up in many more places. Hopefully, we can find a way to enforce it.

macrogreg replied:
agreed

mightyshazam committed:
Merge branch 'feature/unit-tests-2' into feature/github-actions-ci

mightyshazam committed:
removed last system.text.json reference

mightyshazam committed:
added test artifacts for .net framework

mightyshazam committed:
switched to building with dotnet cli

mightyshazam committed:
trying to unify on one build job

mightyshazam committed:
fixed solution file error with nested project

mightyshazam committed:
adjusting trx path

mightyshazam committed:
refactoring artifact path again

mightyshazam committed:
added specific code analysis compilers version to help with 3.1 ci builds

mightyshazam committed:
skipping build and going straight to tests

mightyshazam committed:
added framework arg to test run

mightyshazam committed:
fixed test directory path

mightyshazam committed:
trying single windows job again

mightyshazam committed:
removed additional job

mightyshazam committed:
added paths to exclude

mightyshazam committed:
adding test reporting

mightyshazam committed:
removed test reporter

mightyshazam committed:
updated working directory for test reporter

mightyshazam committed:
changed working directory for test reporter again

mightyshazam committed:
added category for tests

mightyshazam committed:
implemented category filter

macrogreg committed:
Merge pull request #52: Implement all remaining data converters for full parity with other languages.

Implement all remaining data converters for full parity with other languages

macrogreg committed:
Merge pull request #56: Update styling rules and address all styling warnings and errors.

We used to have some stale / outdated styling rules and we had some ignored warnings related to that.
This PR updates the rules and fixes all related errors and warnings.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Common/TestPayload.cs:
```diff
@@ -0,0 +1,68 @@
+            AssertUnnamedCorrectness(3, payload);
+        }
+
+        [Fact]
```
Could we please add a test comparing the two cases outlined here?
https://github.com/temporalio/sdk-dotnet/blob/76ccdb41115dd36d863a35b3f5364820aaea45b8/Src/SDK/Common/Temporal.Serialization/public/PayloadConverter.cs#L197-L212

`Payload.Unnamed<>(new int[] { 100, 200, 300 })` should return a container with 3 items of type int.
`Payload.Unnamed<>(new int[] { 100, 200, 300 })` should return a container with 1 item of type int[] and that item being an array of length 3..

macrogreg replied:
✔️

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Enums/TestWorkflowExecutionStatusExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+using Temporal.Api.Enums.V1;
+using Xunit;
+
+namespace Temporal.Sdk.Common.Tests
```
The namespace of the class being tested is `Temporal.Api.Enums.V1`, so should we call this namespace `Temporal.Api.Enums.V1.Test`?

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Enums/TestWorkflowExecutionStatusExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+using Temporal.Api.Enums.V1;
```
The location of the class being tested is _\Src\SDK\Common\Temporal.Api.Enums.V1\public\WorkflowExecutionStatusExtensions.cs_ 
If the directory structuring suggestion sounds good to you, then this would go into _\Src\**Test\**SDK\Common\Temporal.Api.Enums.V1\public\WorkflowExecutionStatusExtensions.cs_ 

If you prefer a different file placement strategy, I am quite open. The only thing that we need to do is to specify a consistent way how '

macrogreg replied:
File placement discussed and adjusted.

mightyshazam committed:
removed test reporter

macrogreg commented on Src/SDK/Common/Temporal.Serialization/public/ProtobufPayloadConverter.cs:
```diff
@@ -0,0 +1,88 @@
+using System;
```
I think, this file also still has thew line endings problem. :)

macrogreg replied:
Oh no, sorry, this is not the Protobuf Json PayloadConverter :)
Why do you think we need it? All IMessages are already converted to Proto Json. When would this kick in? 
I think our other SDKs do not have the binary Proto conversions except for byte strings, although I only explicitly checked for Java. :) They don't have it.

mightyshazam replied:
I added it. Then I removed it. Then I checked the Java sdk, and I saw it was there, so I went ahead put it back. I found it in the go sdk here https://github.com/temporalio/sdk-go/blob/0fa6692bd57e4db9fee7d15a980a74cea5f8fcae/converter/proto_payload_converter.go and in the java sdk here https://github.com/temporalio/sdk-java/blob/ed3c65f7e3c4a8ed805d1262eee3d17b6dab7b1a/temporal-sdk/src/main/java/io/temporal/common/converter/ProtobufPayloadConverter.java . I think I interpreted it correctly.

macrogreg replied:
You are correct. Wow, looks like I did not sync my local copy of Java SDK since this was added in January :)
https://github.com/temporalio/sdk-java/commit/ed3c65f7e3c4a8ed805d1262eee3d17b6dab7b1a

@lorensr, you made that change? Can you please educate me on the background? I read the PR description, but I am not fully clear under what circumstance we end up doing non-json proto serialization. And, if we do it sometimes, why do we not do it always? I realize, other languages have legacy, but new ones could prefer the binary serialization and only use json if the respective payloads arrive. Wouldn't it be more efficient?

macrogreg replied:
@mightyshazam thanks for discovering it. Let's leave it. Depending on what we learn from @lorensr, we can decide on the priority order between the two serializers.

lorensr replied:
Default converter has this order—ProtoJSON is before ProtoBinary, so JSON is always used. 

https://github.com/temporalio/sdk-go/blob/5e5645f0c550dcf717c095ae32c76a7087d2e985/converter/default_data_converter.go#L28

We may change the default after displaying binary is supported in web & tctl. You can follow https://github.com/temporalio/ui-server/issues/79

macrogreg commented on Src/SDK/WorkflowClient/Temporal.WorkflowClient/internal/TemporalServiceInvoker.cs:
```diff
@@ -1,1061 +1,1061 @@
-﻿using System;
-using System.Diagnostics.CodeAnalysis;
```
Looks like this slipped through when the line endings were corrected :)

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/JsonPayloadConverter.cs:
```diff
@@ -0,0 +1,33 @@
+using Temporal.Serialization;
+using Payload = Temporal.Api.Common.V1.Payload;
+
+namespace Temporal.Sdk.Common.Tests
```
The same comment as above about namespace and placement. :)

mightyshazam replied:
I ultimately removed this class.

mightyshazam committed:
changed endings to \n

mightyshazam committed:
removed some whitespace in the TemporalServiceInvoker class

mightyshazam committed:
renormalized line endings

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Payloads/AbstractUnnamedTest.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace Temporal.Sdk.Common.Tests
+{
+    public abstract class AbstractUnnamedTest
```
The convention elsewhere is `abstract XyzBase`, so perhaps `PayloadContainers.Unnamed.Base`..?

macrogreg replied:
Also, why is this class abstract? It does not have abstract methods. Is that so that xUnit does not instantiate it and only uses it through the subclasses?

mightyshazam replied:
It is so I can exercise common logic across the classes without it being its own test. As I was writing the tests, I noticed I was repeating myself. I debated creating a helper class, but I settled on the abstract class.

macrogreg replied:
ok, cool

✔️

mightyshazam committed:
replaced TemporalServiceInvoker with original file

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Payloads/TestPayloadContainersUnnamedSerializedDataBacked.cs:
```diff
@@ -0,0 +1,66 @@
+            };
+        }
+
+        private class ConvertedClass
```
Would it make sense to reuse [SerializableClass.cs](https://github.com/macrogreg/sdk-dotnet/pull/1/files#diff-5c01bf9ef9746d7e1de9c8d3a548addc2a699b7be8697b4b1e209bc10bb1bf12) here?

mightyshazam replied:
It would. I thought I had removed all instances of `ConvertedClass`, but it looks like one survived.

mightyshazam replied:
The replacement should be complete now.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/SerializableClass.cs:
```diff
@@ -0,0 +1,17 @@
+    internal class SerializableClass
+    {
+        public static SerializableClass Default
+        {
```
I wish I knew how I can teach the style checker communicate this, so sorry about it.

One-liner getters elsewhere are 

```cs
public static SerializableClass Default
{
    get { return new SerializableClass { Name = "Test", Value = 1 }; }
}
```

but this is a nitpick, feel free to ignore. :)

mightyshazam replied:
This was another one from Resharper when the style guide made me change from `=>` syntax to getters.

macrogreg replied:
✔️

mightyshazam committed:
split ci and report

mightyshazam committed:
added branch filter for reporting

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestCatchAllPayloadConverter.cs:
```diff
@@ -0,0 +1,56 @@
+namespace Temporal.Sdk.Common.Tests.Serialization
+{
+    // Currently used for debug/dev
+    public class TestNewtonsoftJsonPayloadConverter
```
File name and class name mismatch

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestCatchAllPayloadConverter.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        [Fact]
+        public void Test_TestNewtonsoftJsonPayloadConverter_Array_Roundtrip()
```
I suggest we should add a test for TryDeserialize for cases where it must return false.
E.g. when there are multiple payloads or for enumerations.

mightyshazam replied:
I agree. My other converter tests do specifically test TryDeserialize/TrySerialize to ensure they return false, but I just wanted to cover the happy paths for the new converters before writing too much code. As you've pointed out in other comments, there are things we should settle on before writing too much code.

macrogreg replied:
agreed

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestCompositePayloadConverter.cs:
```diff
@@ -0,0 +1,66 @@
+            UnnamedContainerPayloadConverter unnamed = new UnnamedContainerPayloadConverter();
+            unnamed.InitDelegates(new[] { new JsonPayloadConverter() });
+            CompositePayloadConverter instance = new CompositePayloadConverter(new IPayloadConverter[]
+            {
```
In this test, why are you using a custom converter list?

mightyshazam replied:
This test was written against the old converter. That was before we had the newtonsoft converter, so it always failed because it went to the catchall converter. That is no longer an issue, so I updated it.

macrogreg replied:
go it, thanks

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestCompositePayloadConverter.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        [Fact]
+        public void Test_CompositePayloadConverter_Catchall_Roundtrip()
```
Is this not a duplicate of  `Test_CompositePayloadConverter_Void_Roundtrip`?

mightyshazam replied:
It is. I was copied and pasted all of the tests from the first. Then I realized the catchall converter wasn't designed to roundtrip data and I never deleted the test. It's gone now.

mightyshazam replied:
I have as a mental todo to add roundtrip tests for more types, though, to ensure the composite converter is working properly.

macrogreg replied:
cool
✔️

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestNullPayloadConverter.cs:
```diff
@@ -0,0 +1,78 @@
+            ByteString bs = null;
+            NullPayloadConverter instance = new NullPayloadConverter();
+            ByteString b = PayloadConverter.GetOrCreateBytes(
+                NullPayloadConverter.PayloadMetadataEncodingValue,
```
I suggest, rather than referring to `NullPayloadConverter.PayloadMetadataEncodingValue`, we can use a `const string NullPayloadConverterMetadataEncodingValue = "binary/null"` or so. that way, if `NullPayloadConverter.PayloadMetadataEncodingValue` was to change, this test would catch it. 
The benefit that that we need to ansure roundtrip not only with .NET, but also with other languages, so MetadataEncodingValue is, essentially, fixed.

mightyshazam replied:
I agree. Both the go and java sdk use a separate class of constants and reference them in the appropriate converter. Here is the java example https://github.com/temporalio/sdk-java/blob/ed3c65f7e3c4a8ed805d1262eee3d17b6dab7b1a/temporal-sdk/src/main/java/io/temporal/common/converter/EncodingKeys.java
In this case, I was writing a roundtrip test, but it was the first test I wrote, so I didn't have a clear understanding of the code. If I were to do it again, I would just use the converter to encode the data and make sure it deserializes. As an additional step, we could test the metadata to make sure it's correct.

macrogreg replied:
Agree to all. Changing that one aspect might be a low-hanging fruit. :)
I know java has one central place for all keys. When I wrote the .NET equivalent I tried to write the converters like a user would, if they were writing a custom converter. And from that perspective, putting the constant with the converter appeared more fitting.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestNullPayloadConverter.cs:
```diff
@@ -0,0 +1,78 @@
+        public void Test_NullPayloadConverter_TryDeserialize_Nonnullable_Type()
+        {
+            NullPayloadConverter instance = new NullPayloadConverter();
+            Assert.False(instance.TryDeserialize(
```
```cs
Assert.False(instance.TryDeserialize(new Payloads { Payloads_ = { new Payload() }, },
                                     out int _));
```

macrogreg replied:
✔️

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestNullPayloadConverter.cs:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        [Fact]
+        public void Test_NullPayloadConverter_TrySerialize_Null()
```
Perhaps also test a nullable `int?`?

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestProtobufJsonPayloadConverter.cs:
```diff
@@ -0,0 +1,33 @@
+    public class TestProtobufJsonPayloadConverter
+    {
+        [Fact]
+        public void Test_ProtobufPayloadConverter_ByteString_Roundtrip()
```
Why ByteString?

mightyshazam replied:
That shouldn't say `ByteString`. It should say `Proto`

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestRawMemoryPayloadConverter.cs:
```diff
@@ -0,0 +1,95 @@
+        [Fact]
+        public void Test_RawMemoryPayloadConverter_ReadonlyMemory_Roundtrip()
+        {
+            Random r = new();
```
In this case it does not matter, but generally, I would prefer not to use random input int tests, because if it fails, it is not reproducible. We could work around it by specifying a constant byte array or by specifying a constant seed to the Random. (same for other test cases)

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestRawMemoryPayloadConverter.cs:
```diff
@@ -0,0 +1,95 @@
+
+        [Fact]
+        // TODO: Determine whether this should successfully roundtrip
+        public void Test_RawMemoryPayloadConverter_ByteArray_Roundtrip()
```
I really like that we have this test case. The Try methods should return false for the same reason as we do not convert enumerable in other converters: not clear whether user meant one array-arg or multiple byte args.
If people have a byte array to encode as raw, they can use [Payload.Raw(..)](url) at the call site. In the future, if there are user requests, we can add more `Payload.Raw(..)` overloads to support other primitive type arrays.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestUnnamedContainerPayloadConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            UnnamedContainerPayloadConverter instance = new UnnamedContainerPayloadConverter();
+            Payloads p = new Payloads();
+            Assert.False(instance.TrySerialize<string>(null, p));
+            Assert.Empty(p.Payloads_);
```
I suggest adding this assert to this test case:

```cs
Assert.False(instance.TrySerialize<PayloadContainers.UnnamedInstanceBacked<object>>(null, p));
```

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestUnnamedContainerPayloadConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            Assert.NotEmpty(p.Payloads_);
+            Assert.True(instance.TryDeserialize(p, out PayloadContainers.Unnamed.SerializedDataBacked cl));
+            Assert.NotNull(cl);
+            SerializableClass deserializedData = cl.GetValue<SerializableClass>(0);
```
Let's also check that `cl.PayloadConverter` is set to `instance`.

mightyshazam replied:
I added the check. `cl.PayloadConverter` is actually equal to `CompositePayloadConverter`. This is correct behavior, but it seems weird until you realize the `UnnamedContainerPayloadConverter` uses the `CompositePayloadConverter` behind the scenes.

macrogreg replied:
I modified that test case to account for this in a later PR.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestUnnamedContainerPayloadConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            Assert.NotNull(cl);
+            SerializableClass deserializedData = cl.GetValue<SerializableClass>(0);
+            Assert.Equal("test", deserializedData.Name);
+            Assert.Equal(2, deserializedData.Value);
```
This is a great test case. I would like to add an additional to check that the converter embedded in the container is invoked when it should be:

first time you call `instance.TrySerialize(data, p)`, the `JsonPayloadConverter `you set in line `instance.InitDelegates(new[] { new JsonPayloadConverter() });` should be invoked. Since that's a test converter, we con modify it to add a respective query method or so.

After we have `cl`, use `instance` to serialize `cl` again. Now the `TrySerialize` method of the embedded `JsonPayloadConverter` should not be called, because the container serializer should notice the same underlying serializer and just round-trip the backing data.

We can also validate a match for `cl.SerializedData`.

Lastly construct a new `instance2` and apply it to the came `cl`. This time the underlying `JsonPayloadConverter` converter should be called again, since we are not using the same converter and cannot guarantee that using raw payload is valid.

We can also chat about that mechanism if you like. Perhaps it's too complex? Should we have comments in such a complex test?

macrogreg replied:
I modified that test case to account for most of this in a later PR, and added a comment about outstanding @ToDo.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestUnnamedContainerPayloadConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            Assert.True(instance.TrySerialize(data, p));
+            Assert.NotEmpty(p.Payloads_);
+            Assert.False(instance.TryDeserialize(p, out PayloadContainers.Unnamed.InstanceBacked<string> _));
+            Assert.True(instance.TryDeserialize(p, out PayloadContainers.Unnamed.SerializedDataBacked cl));
```
I would also test for
`Assert.True(instance.TryDeserialize(p, out PayloadContainers.IUnnamed cl2));`
and assert that `cl2 is not null && cl2 is PayloadContainers.Unnamed.SerializedDataBacked`

mightyshazam replied:
Added.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestUnnamedContainerPayloadConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            Assert.True(instance.TryDeserialize(p, out PayloadContainers.Unnamed.SerializedDataBacked cl));
+            Assert.NotEmpty(cl);
+            Assert.True(cl.TryGetValue(0, out string val));
+            Assert.Equal("hello", val);
```
Do you think it makes sense to test the other cases you have in TestPayload or is that an overkill?

mightyshazam replied:
It may be overkill at the moment.

macrogreg commented on Src/Tests/Temporal.Sdk.Common.Tests/Serialization/TestUnnamedContainerPayloadConverter.cs:
```diff
@@ -0,0 +1,74 @@
+            Assert.True(instance.TrySerialize(data, p));
+            Assert.Empty(p.Payloads_);
+        }
+    }
```
I think it might be useful to also have a nested test case with a dataset along these lines:
https://github.com/temporalio/sdk-dotnet/blob/3054a2282b2eabf86d94fc970db6d4da55f8746d/Src/Demos/AdHocScenarios/Temporal.Demos.AdHocScenarios/AdHocClientInvocations.cs#L99-L108

macrogreg reviewed (commented):
Since this PR is a placeholder, I'll provide the summary comments in the PR it actually represents:
https://github.com/temporalio/sdk-dotnet/pull/53

macrogreg commented on Src/SDK/Common/Temporal.Serialization/public/IPayloadCodec.cs:
```diff
@@ -1,12 +1,12 @@
-﻿using System.Threading;
-using System.Threading.Tasks;
```
Oh, I think the line endings slipped through here as well. :)

mightyshazam replied:
I got this one.

mightyshazam committed:
replaced IPayloadCodec with original file

macrogreg committed:
Merge pull request #53 from mightyshazam/feature/unit-tests-2

Added unit tests again with autocrlf enabled
When you merge, could you please modify GitHub's default suggestion for the merge commit name from 
`Merge pull request #53 from mightyshazam/feature/unit-tests-2`
to
`Merge pull request #53: Added unit tests again with autocrlf enabled`

The pattern is
`Merge pull request #xxx` + `: ` + _name of the PR_

mightyshazam committed:
updated namespaces in TestPayload to shorten type names

mightyshazam committed:
minor code review refactor

mightyshazam committed:
Code review changes
- Renamed checkout task
- Renamed job in ci build
- Updated badge repository reference in readme

mightyshazam committed:
Merge remote-tracking branch 'upstream/master' into feature/github-actions-ci

macrogreg committed:
Initial scaffolding for integration tests.

macrogreg committed:
SimpleClientInvocations works as E2E integration Test based in TemporalLite

macrogreg committed:
TemporalClientTest scaffolding complete and working on test cases (builds and runs).

macrogreg committed:
StartWorkflowAsync tests.

mightyshazam committed:
more code review changes: renamed testresults folder and made categories assembly-wide

mightyshazam committed:
added a test to show difference between unnamed<int> and unnamed<int[]> when given array arguments

mightyshazam committed:
added more code review changes

macrogreg committed:
Merge pull request #59: Set up GitHib action workflows for CI (part 1)

### Summary

This PR sets up basic CI gates based on the unit tests so far.

 * **This PR is a Part 1 of the work on CI. It will be completed in the next PR.**
 
 * Note: I set up the PR, but all the work here is by @mightyshazam.


### Related work items

* https://github.com/temporalio/sdk-dotnet/issues/14

macrogreg committed:
Adjust GH Actions scripts names and add them to the solution. (#63)

mightyshazam committed:
Merge remote-tracking branch 'upstream/master' into feature/unit-tests-2

macrogreg committed:
Rename `Tests` to `Test`. (#64)

macrogreg committed:
Make GH workflows more granular and target additional branches. (#66)

* Adjust GH Actions scripts names and add them to the solution.

* Make GH workflows more granular and target additional branches.

macrogreg committed:
Merge ongoing work on CI scripts (#68)

macrogreg committed:
Merge ongoing work on CI scripts (#69)

macrogreg committed:
Merge ongoing work on CI scripts (#70)

macrogreg committed:
Improve name for CI workflows. (#73)

macrogreg committed:
`continue-on-error: true` for Test Report Generators. (#74)

macrogreg committed:
Debug info in labels for report generators. (#75)

macrogreg committed:
No bin upload for Debug; Skip non-Win Int Test report. (#76)

macrogreg committed:
Merge Unit tests and Int tests with CI scripts.

macrogreg committed:
Move all tests to structured file locations.

macrogreg committed:
Merge pull request #60: Add automatic integration tests

### Summary

* Add automatic integration tests, including requires scaffolding and several actual tests.
* Move Unit tests to expected file locations.

### Related work items

* https://github.com/temporalio/sdk-dotnet/issues/13

### Details

This PR adds the first set of integration tests. Specifically, this includes:
* Set up the general test project structure.
* Establish testing frameworks.
* Enable automatic deployment of TemporalLite as test server for auto-integration tests (Windows only).
* Add an automatic E2E integration test (based on existing manual tests).
* Add automatic per method integration tests for a subset of client methods.

### Follow-ups

Tests need to be incorporated into the CI in subsequent PRs.

macrogreg committed:
Add test projects to respective automation solutions.

mightyshazam committed:
Merge remote-tracking branch 'upstream/master' into feature/unit-tests-2

macrogreg committed:
Fix path when cloning build tools repo.

macrogreg committed:
Execute Int Tests sequentially to avoid contention on server proc.

macrogreg committed:
Fix broken Serialization tests.

macrogreg committed:
Merge pull request #77: Activate Unit and Integration Tests in CI

### Summary

This change adds all respective test projects that exist so far to the respective automation solutions.
The CI scripts will pick this up and the tests will become active as CI validation gates.

### Related work items

* https://github.com/temporalio/sdk-dotnet/issues/13
* https://github.com/temporalio/sdk-dotnet/issues/14

mightyshazam committed:
Merge branch 'master' into feature/unit-tests-2
